[mono] Crash when calling an interface method on a struct through a generic default interface method

Reported by @Jorenkv at https://github.com/mono/mono/issues/21206 but also happens in .NET 6

```csharp
using System;

class Program {
    static void Main() {
        (new Foo() as IFoo).NoCrash(); // Prints "Hello!" as expected
        (new Foo() as IFoo).Crash<int>(); // Crashes
    }
}

interface IFoo {
    void Crash<T>() {
        Bla();
    }

    void NoCrash() {
        Bla();
    }

    void Bla();
}

struct Foo: IFoo {
    public void Bla() {
        Console.WriteLine("Hello!");
    }
}
```

Expected result: prints Hello! twice

Actual result:
```
Hello!

=================================================================
	Native Crash Reporting
=================================================================
Got a SIGSEGV while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries
used by your application.
=================================================================

=================================================================
	Native stacktrace:
=================================================================
	0x1040c0184 - /Users/alklig/work/dotnet-runtime/runtime/artifacts/bin/HelloWorld/x64/Release/osx-x64/publish/libcoreclr.dylib : mono_dump_native_crash_info
	0x1040627ae - /Users/alklig/work/dotnet-runtime/runtime/artifacts/bin/HelloWorld/x64/Release/osx-x64/publish/libcoreclr.dylib : mono_handle_native_crash
	0x103fbd30b - /Users/alklig/work/dotnet-runtime/runtime/artifacts/bin/HelloWorld/x64/Release/osx-x64/publish/libcoreclr.dylib : mono_sigsegv_signal_handler_debug
	0x7fff20486d7d - /usr/lib/system/libsystem_platform.dylib : _sigtramp
	0x7fff2026af0a - /usr/lib/system/libsystem_malloc.dylib : tiny_malloc_from_free_list
	0x103fc0579 - /Users/alklig/work/dotnet-runtime/runtime/artifacts/bin/HelloWorld/x64/Release/osx-x64/publish/libcoreclr.dylib : mono_jit_runtime_invoke
	0x103edea48 - /Users/alklig/work/dotnet-runtime/runtime/artifacts/bin/HelloWorld/x64/Release/osx-x64/publish/libcoreclr.dylib : mono_runtime_invoke_checked
	0x103ee688c - /Users/alklig/work/dotnet-runtime/runtime/artifacts/bin/HelloWorld/x64/Release/osx-x64/publish/libcoreclr.dylib : do_exec_main_checked
	0x10401d132 - /Users/alklig/work/dotnet-runtime/runtime/artifacts/bin/HelloWorld/x64/Release/osx-x64/publish/libcoreclr.dylib : mono_jit_exec
	0x10402076a - /Users/alklig/work/dotnet-runtime/runtime/artifacts/bin/HelloWorld/x64/Release/osx-x64/publish/libcoreclr.dylib : mono_main
	0x10409b5d3 - /Users/alklig/work/dotnet-runtime/runtime/artifacts/bin/HelloWorld/x64/Release/osx-x64/publish/libcoreclr.dylib : monovm_execute_assembly
	0x103e144e1 - /Users/alklig/work/dotnet-runtime/runtime/artifacts/bin/HelloWorld/x64/Release/osx-x64/publish/libhostpolicy.dylib : _Z19run_app_for_contextRK20hostpolicy_context_tiPPKc
	0x103e15401 - /Users/alklig/work/dotnet-runtime/runtime/artifacts/bin/HelloWorld/x64/Release/osx-x64/publish/libhostpolicy.dylib : corehost_main
	0x103da1bce - /Users/alklig/work/dotnet-runtime/runtime/artifacts/bin/HelloWorld/x64/Release/osx-x64/publish/libhostfxr.dylib : _ZN10fx_muxer_t24handle_exec_host_commandERKNSt3__112basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEERK19host_startup_info_tS8_RKNS0_13unordered_mapI13known_optionsNS0_6vectorIS6_NS4_IS6_EEEE18known_options_hashNS0_8equal_toISD_EENS4_INS0_4pairIKSD
	0x103da0a4b - /Users/alklig/work/dotnet-runtime/runtime/artifacts/bin/HelloWorld/x64/Release/osx-x64/publish/libhostfxr.dylib : _ZN10fx_muxer_t7executeENSt3__112basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEEiPPKcRK19host_startup_info_tPciPi
	0x103d9d4d8 - /Users/alklig/work/dotnet-runtime/runtime/artifacts/bin/HelloWorld/x64/Release/osx-x64/publish/libhostfxr.dylib : hostfxr_main_startupinfo
	0x103d54795 - /Users/alklig/work/dotnet-runtime/runtime/src/mono/sample/HelloWorld/../../../../artifacts/bin/HelloWorld/x64/Release/osx-x64/publish/HelloWorld : _Z9exe_startiPPKc
	0x103d549af - /Users/alklig/work/dotnet-runtime/runtime/src/mono/sample/HelloWorld/../../../../artifacts/bin/HelloWorld/x64/Release/osx-x64/publish/HelloWorld : main
	0x7fff2045cf3d - /usr/lib/system/libdyld.dylib : start

=================================================================
	External Debugger Dump:
=================================================================

=================================================================
	Basic Fault Address Reporting
=================================================================
Memory around native instruction pointer (0x107c4704b):0x107c4703b  49 8b ff 49 8b 07 49 ba e0 89 80 8a e7 7f 00 00  I..I..I.........
0x107c4704b  ff 50 88 4c 8b 3c 24 48 83 c4 08 c3 00 00 00 00  .P.L.<$H........
0x107c4705b  00 00 00 00 00 48 83 c7 10 48 b8 10 70 c4 07 01  .....H...H..p...
0x107c4706b  00 00 00 ff e0 00 00 00 00 00 00 00 00 00 00 00  ................

=================================================================
	Managed Stacktrace:
=================================================================
	  at IFoo:Crash <0x0003b>
	  at Program:RunTwo <0x00064>
	  at Program:Main <0x00042>
	  at System.Object:runtime_invoke_void <0x000a5>
=================================================================
make: *** [run] Abort trap: 6
```

Does not happen if `Foo` is a `class` not a `struct`

--- a/src/mono/mono/mini/mini-trampolines.c
+++ b/src/mono/mono/mini/mini-trampolines.c
@@ -632,11 +632,13 @@ common_call_trampoline (host_mgreg_t *regs, guint8 *code, MonoMethod *m, MonoVTa
 		return NULL;
 
 	if (generic_virtual || variant_iface) {
-		if (m_class_is_valuetype (vt->klass)) /*FIXME is this required variant iface?*/
+		if (m_class_is_valuetype (vt->klass)  && !mini_method_is_default_method (m)) /*FIXME is this required variant iface?*/
 			need_unbox_tramp = TRUE;
 	} else if (orig_vtable_slot) {
-		if (m_class_is_valuetype (m->klass))
+		if (m_class_is_valuetype (m->klass)) {
+			g_assert (!mini_method_is_default_method (m));
 			need_unbox_tramp = TRUE;
+		}
 	}
 
 	addr = mini_add_method_trampoline (m, compiled_method, need_rgctx_tramp, need_unbox_tramp);

--- a/src/tests/Loader/classloader/DefaultInterfaceMethods/regressions/github58394.cs
+++ b/src/tests/Loader/classloader/DefaultInterfaceMethods/regressions/github58394.cs
@@ -1,0 +1,47 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.CompilerServices;
+
+namespace GenericDimValuetypeBug
+{
+    class Program
+    {
+        static int Main()
+        {
+            if (RunOne() != 17)
+                return 1;
+            if (RunTwo() != 23)
+                return 2;
+            return 100;
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public static int RunOne()
+        {
+            return (new Foo() { x = 17 } as IFoo).NoCrash();
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public static int RunTwo()
+        {
+            return (new Foo() { x = 23 } as IFoo).Crash<int>();
+        }
+    }
+
+    interface IFoo
+    {
+        int Crash<T>() => Bla();
+
+        int NoCrash() => Bla();
+
+        int Bla();
+    }
+
+    struct Foo: IFoo
+    {
+        public int x;
+        public int Bla() => x;
+    }
+}

--- a/src/tests/Loader/classloader/DefaultInterfaceMethods/regressions/github58394.csproj
+++ b/src/tests/Loader/classloader/DefaultInterfaceMethods/regressions/github58394.csproj
@@ -1,0 +1,11 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <OutputType>Exe</OutputType>
+    <CLRTestKind>BuildAndRun</CLRTestKind>
+    <CLRTestPriority>0</CLRTestPriority>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="$(MSBuildProjectName).cs" />
+  </ItemGroup>
+</Project>

